Variable mutability in extracted function

Now we don't care about variable mutability in "extract function" refactoring and we can generate incorrect code in some cases

```rust
fn foo() {
    let mut vec = vec![1, 2, 3];
    <selection>vec.push(1);</selection>
}
```

**Expected:**
```rust
fn foo() {
    let mut vec = vec![1, 2, 3];
    bar(vec);
}

fn bar(mut vec: Vec<i32>) {
    vec.push(1);
}
```

or even

```rust
fn foo() {
    let mut vec = vec![1, 2, 3];
    bar(&mut vec);
}

fn bar(vec: &mut Vec<i32>) {
    vec.push(1);
}
```

**Actual:**
```rust
fn foo() {
    let mut vec = vec![1, 2, 3];
    bar(vec);
}

fn bar(vec: Vec<i32>) {
    vec.push(1); // error[E0596]: cannot borrow immutable argument `vec` as mutable
}
```

--- a/src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt
+++ b/src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionConfig.kt
@@ -16,6 +16,7 @@ import org.rust.lang.core.psi.RsFunction
 import org.rust.lang.core.psi.RsPatBinding
 import org.rust.lang.core.psi.RsTypeParameter
 import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.types.ty.Mutability
 import org.rust.lang.core.types.ty.Ty
 import org.rust.lang.core.types.ty.TyTuple
 import org.rust.lang.core.types.type
@@ -39,10 +40,10 @@ class ReturnValue(val expression: String?, val type: Ty) {
     }
 }
 
-class Parameter(val name: String, val type: Ty? = null) {
+class Parameter(val name: String, val type: Ty? = null, val mutability: Mutability = Mutability.IMMUTABLE) {
     companion object {
         fun direct(value: RsPatBinding): Parameter {
-            return Parameter(value.referenceName, value.type)
+            return Parameter(value.referenceName, value.type, value.mutability)
         }
 
         fun self(value: String): Parameter {
@@ -51,10 +52,15 @@ class Parameter(val name: String, val type: Ty? = null) {
     }
 
     val text: String
-        get() = if (type != null) {
-            "$name: ${type.insertionSafeText}"
-        } else {
-            name
+        get() = buildString {
+            if (type != null) {
+                if (mutability.isMut) {
+                    append("mut ")
+                }
+                append("$name: ${type.insertionSafeText}")
+            } else {
+                append(name)
+            }
         }
 }
 

--- a/src/test/kotlin/org/rust/lang/refactoring/RsExtractFunctionTest.kt
+++ b/src/test/kotlin/org/rust/lang/refactoring/RsExtractFunctionTest.kt
@@ -118,6 +118,40 @@ class RsExtractFunctionTest : RsTestBase() {
         false,
         "foo")
 
+    fun `test extract input parameter from method`() = doTest("""
+            fn foo(a: i32) {
+                <selection>println!("{}", a);</selection>
+            }
+        """, """
+            fn foo(a: i32) {
+                bar(a);
+            }
+
+            fn bar(a: i32) {
+                println!("{}", a);
+            }
+        """,
+        false,
+        "bar")
+
+    fun `test extract input parameter with mutability`() = doTest("""
+            fn main() {
+                let mut vec = vec![1, 2, 3];
+                <selection>vec.push(1);</selection>
+            }
+        """, """
+            fn main() {
+                let mut vec = vec![1, 2, 3];
+                foo(vec);
+            }
+
+            fn foo(mut vec: Vec<i32>) {
+                vec.push(1);
+            }
+        """,
+        false,
+        "foo")
+
     fun `test extract two input parameter`() = doTest("""
             fn main() {
                 let bar = 10i32;

